### PR TITLE
Allow DateSpan to set a max number of days

### PR DESCRIPTION
**Pull request by kaapstorm**
This works towards allowing DailyFormStatsReport to set a maximum date span, to avoid errors caused by massive DataTable async request headers.

Context: [FB 185304](http://manage.dimagi.com/default.asp?185304)

@snopoke, @czue, cc @millerdev

### CODE REVIEW

**Comment by millerdev** [comment]
:+1:

**Comment by czue** [comment]
nice tests!

File: dimagi/utils/dates.py [hunk 1]
```diff
@@ -136,6 +136,7 @@ class DateSpan(object):
     format = None
     inclusive = True
     is_default = False
+    max_days = -1
```
Comment (czue) [inline dimagi/utils/dates.py]: nit picky but I think `None` might be a clearer "not set" default.

File: dimagi/utils/dates.py [hunk 2]
```diff
@@ -303,6 +306,11 @@ def get_validation_reason(self):
             return "You can't have an end date of %s after start date of %s" % (self.enddate, self.startdate)
         elif self.startdate < datetime.datetime(1900, 01, 01) or self.enddate < datetime.datetime(1900, 01, 01):
             return "You can't use dates earlier than the year 1900"
+        elif self.max_days >= 0:
+            delta = self.enddate - self.startdate
+            if delta.days > self.max_days:
+                return "You are limited to a span of {} days, but this date range spans {} days".format(
```
Comment (czue) [inline dimagi/utils/dates.py]: if these are intended to be user facing i wonder if they should be translated (i see above the others aren't so not blocking)

**Comment by czue** [comment]
was there a reason you didn't put max_days in the constructor? I think I would have found that more intuitive to use

**Comment by czue** [comment]
(no blockers on my end)

File: dimagi/utils/dates.py (hunk 1 above)
Comment (kaapstorm) [inline dimagi/utils/dates.py]: I agree. Labelling "Don't merge" while I change that.

**Comment by kaapstorm** [comment]
(Actually, I put max_days in the constructor, and then took it out because I found I wasn't using it that way and I figured I'd keep my change smaller. But maybe my first instinct was right, if you think you'd expect it to be there. I'll put it back.)

**Comment by kaapstorm** [comment]
Failed locally too, and I was hoping it was just my environment. 
```
  File "/home/travis/build/dimagi/dimagi-utils/dimagi/utils/dates.py", line 3, in <module>
    from django.utils.translation import ugettext_lazy as _
ImportError: No module named utils.translation
```
Any ideas why the import isn't working? Django is a listed requirement, and installed, and my IDE finds `utils.translation` no problem.

**Comment by snopoke** [comment]
I think that can happen if there are circular imports

**Comment by snopoke** [comment]
Or if it tries to load translations before they are initialized

File: dimagi/utils/dates.py [hunk 3]
```diff
@@ -303,6 +308,11 @@ def get_validation_reason(self):
             return "You can't have an end date of %s after start date of %s" % (self.enddate, self.startdate)
         elif self.startdate < datetime.datetime(1900, 01, 01) or self.enddate < datetime.datetime(1900, 01, 01):
             return "You can't use dates earlier than the year 1900"
+        elif self.max_days >= 0:
```
Comment (czue) [inline dimagi/utils/dates.py]: since we changed the default to `None` do we also have to add that check here? otherwise I think this might crash.

would be good to add a test-case for this either way (and especially if it does fail)

File: dimagi/utils/tests/dates.py [hunk 4]
```diff
@@ -108,3 +108,41 @@ def dummy(req):
...
+        self.assertEqual(datespan.get_validation_reason(),
+                         "You can't have an end date of %s after start date of %s" % (enddate, startdate))
+
+    def test_wrong_century(self):
+        datespan = DateSpan(datetime(1815, 01, 01), datetime(1815, 02, 01))
+        self.assertFalse(datespan.is_valid())
+        self.assertEqual(datespan.get_validation_reason(), "You can't use dates earlier than the year 1900")
+
+    def test_datespan_ok(self):
+        datespan = DateSpan(datetime(2015, 01, 01), datetime(2015, 04, 01))
```
Comment (czue) [inline dimagi/utils/tests/dates.py]: tests should switch to using the constructor, no?

File: dimagi/utils/dates.py (hunk 3 above)
Comment (kaapstorm) [inline dimagi/utils/dates.py]: Thanks. I missed that one.

File: dimagi/utils/tests/dates.py (hunk 4 above)
Comment (kaapstorm) [inline dimagi/utils/tests/dates.py]: :+1: I left them to mimic the way max_days is [set in HQ](https://github.com/dimagi/commcare-hq/blob/limit_date_range/corehq/apps/reports/standard/__init__.py#L318-318), but this is neater, and amounts to the same thing. Changing.

Comment (czue) [inline dimagi/utils/tests/dates.py]: cool, I think it also makes sense to change in HQ

**Comment by kaapstorm** [comment]
Translations added. (The problem was that `dimagi.utils.django` was being imported instead of `django`. Using `from __future__ import absolute_imports` fixed it.)

File: dimagi/locale/en_US/LC_MESSAGES/django.po [hunk 5]
```diff
@@ -0,0 +1,38 @@
+# SOME DESCRIPTIVE TITLE.
```
Comment (czue) [inline dimagi/locale/en_US/LC_MESSAGES/django.po]: i don't think we should add these (the HQ translations should pick them up)

File: settings.py [hunk 6]
```diff
@@ -17,6 +17,10 @@
     }
 }
 
+MIDDLEWARE_CLASSES = (
```
Comment (czue) [inline settings.py]: why was this added?

**Comment by czue** [comment]
everything looks good now but i think can remove at least the translations and possibly the settings additions? other than that :+1: from my side. thanks for following up on feedback and sorry this got more complicated than expected!

File: dimagi/locale/en_US/LC_MESSAGES/django.po (hunk 5 above)
Comment (kaapstorm) [inline dimagi/locale/en_US/LC_MESSAGES/django.po]: Cool. I wasn't sure whether to add this for when dimagi-utils is used independently of HQ, so I made sure to put it in a separate commit. I'll drop it.

File: settings.py (hunk 6 above)
Comment (kaapstorm) [inline settings.py]: It's to avoid this warning: 
```
System check identified some issues:

WARNINGS:
?: (1_7.W001) MIDDLEWARE_CLASSES is not set.
	HINT: Django 1.7 changed the global defaults for the MIDDLEWARE_CLASSES. 
django.contrib.sessions.middleware.SessionMiddleware, 
django.contrib.auth.middleware.AuthenticationMiddleware, and 
django.contrib.messages.middleware.MessageMiddleware were removed from the defaults. 
If your project needs these middleware then you should configure this setting.
```

It just explicitly sets MIDDLEWARE_CLASSES to their (Django 1.7) default value.

It's not relevant to this bug -- it was just annoying. I'll take it out this PR, and we can include it when we upgrade dimagi-utils to Django 1.7.

Comment (czue) [inline settings.py]: (seems fine to keep as long as there was a reason - though maybe a comment would be good)

either way i see you've already removed it which is also fine
